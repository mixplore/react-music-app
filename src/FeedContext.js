import React, { useState, createContext, useEffect } from "react";
import { getTopAlbums, getTopSongs } from "./api/requests";
import { getParsedRecords } from "./utils/dataParsers";

export const FeedContext = createContext();

export const FeedProvider = (props) => {
  const [state, setState] = useState({
      feedData: {},
      feedType: 'topalbums',
      feedLimit: 100,
      isLoading: false,
  });

  useEffect(() => {
    async function getData() {
        setState(prevState => ({
          ...prevState,
          isLoading: true
      }));
      const data =
        state.feedType === "topalbums"
          ? await getTopAlbums(state.feedLimit)
          : await getTopSongs(state.feedLimit);
      const feedData = getParsedRecords(data, state.feedType);
      setState(prevState => ({
          ...prevState,
          feedData,
          isLoading: false
      }));
    }
    getData();
  }, [state.feedType]);

  return (
    <FeedContext.Provider value={[state, setState]}>
      {props.children}
    </FeedContext.Provider>
  );
};
