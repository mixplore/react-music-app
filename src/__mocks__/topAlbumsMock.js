export default {
  feed: {
    author: {
      name: {
        label: "iTunes Store",
      },
      uri: {
        label: "http://www.apple.com/itunes/",
      },
    },
    entry: [
      {
        "im:name": {
          label: "Medicine At Midnight",
        },
        "im:image": [
          {
            label:
              "https://is1-ssl.mzstatic.com/image/thumb/Music114/v4/0b/f4/7b/0bf47b0c-f4d9-5766-2de2-88df8b5bc4a1/886448658054.jpg/55x55bb.png",
            attributes: {
              height: "55",
            },
          },
          {
            label:
              "https://is3-ssl.mzstatic.com/image/thumb/Music114/v4/0b/f4/7b/0bf47b0c-f4d9-5766-2de2-88df8b5bc4a1/886448658054.jpg/60x60bb.png",
            attributes: {
              height: "60",
            },
          },
          {
            label:
              "https://is5-ssl.mzstatic.com/image/thumb/Music114/v4/0b/f4/7b/0bf47b0c-f4d9-5766-2de2-88df8b5bc4a1/886448658054.jpg/170x170bb.png",
            attributes: {
              height: "170",
            },
          },
        ],
        "im:itemCount": {
          label: "9",
        },
        "im:price": {
          label: "$9.99",
          attributes: {
            amount: "9.99000",
            currency: "USD",
          },
        },
        "im:contentType": {
          "im:contentType": {
            attributes: {
              term: "Album",
              label: "Album",
            },
          },
          attributes: {
            term: "Music",
            label: "Music",
          },
        },
        rights: {
          label:
            "℗ 2021 Roswell Records, Inc., under exclusive license to RCA Records",
        },
        title: {
          label: "Medicine At Midnight - Foo Fighters",
        },
        link: {
          attributes: {
            rel: "alternate",
            type: "text/html",
            href:
              "https://music.apple.com/us/album/medicine-at-midnight/1538671506?uo=2",
          },
        },
        id: {
          label:
            "https://music.apple.com/us/album/medicine-at-midnight/1538671506?uo=2",
          attributes: {
            "im:id": "1538671506",
          },
        },
        "im:artist": {
          label: "Foo Fighters",
          attributes: {
            href: "https://music.apple.com/us/artist/foo-fighters/6906197?uo=2",
          },
        },
        category: {
          attributes: {
            "im:id": "21",
            term: "Rock",
            scheme: "https://music.apple.com/us/genre/music-rock/id21?uo=2",
            label: "Rock",
          },
        },
        "im:releaseDate": {
          label: "2021-02-05T00:00:00-07:00",
          attributes: {
            label: "February 5, 2021",
          },
        },
      },
      {
        "im:name": {
          label: "The Lucky Ones",
        },
        "im:image": [
          {
            label:
              "https://is2-ssl.mzstatic.com/image/thumb/Music124/v4/c3/c6/95/c3c69540-c9c7-6fe2-15b4-c05064f1ddb7/886448750680.jpg/55x55bb.png",
            attributes: {
              height: "55",
            },
          },
          {
            label:
              "https://is4-ssl.mzstatic.com/image/thumb/Music124/v4/c3/c6/95/c3c69540-c9c7-6fe2-15b4-c05064f1ddb7/886448750680.jpg/60x60bb.png",
            attributes: {
              height: "60",
            },
          },
          {
            label:
              "https://is2-ssl.mzstatic.com/image/thumb/Music124/v4/c3/c6/95/c3c69540-c9c7-6fe2-15b4-c05064f1ddb7/886448750680.jpg/170x170bb.png",
            attributes: {
              height: "170",
            },
          },
        ],
        "im:itemCount": {
          label: "11",
        },
        "im:price": {
          label: "$9.99",
          attributes: {
            amount: "9.99000",
            currency: "USD",
          },
        },
        "im:contentType": {
          "im:contentType": {
            attributes: {
              term: "Album",
              label: "Album",
            },
          },
          attributes: {
            term: "Music",
            label: "Music",
          },
        },
        rights: {
          label: "℗ 2021 RCA Records, a division of Sony Music Entertainment",
        },
        title: {
          label: "The Lucky Ones - Pentatonix",
        },
        link: {
          attributes: {
            rel: "alternate",
            type: "text/html",
            href:
              "https://music.apple.com/us/album/the-lucky-ones/1535661390?uo=2",
          },
        },
        id: {
          label:
            "https://music.apple.com/us/album/the-lucky-ones/1535661390?uo=2",
          attributes: {
            "im:id": "1535661390",
          },
        },
        "im:artist": {
          label: "Pentatonix",
          attributes: {
            href: "https://music.apple.com/us/artist/pentatonix/466047278?uo=2",
          },
        },
        category: {
          attributes: {
            "im:id": "14",
            term: "Pop",
            scheme: "https://music.apple.com/us/genre/music-pop/id14?uo=2",
            label: "Pop",
          },
        },
        "im:releaseDate": {
          label: "2021-02-12T00:00:00-07:00",
          attributes: {
            label: "February 12, 2021",
          },
        },
      },
      {
        "im:name": {
          label: "Life Rolls On",
        },
        "im:image": [
          {
            label:
              "https://is1-ssl.mzstatic.com/image/thumb/Music114/v4/4c/31/ba/4c31ba09-c78d-897e-6249-f25afe489382/20BMR0002146.rgb.jpg/55x55bb.png",
            attributes: {
              height: "55",
            },
          },
          {
            label:
              "https://is3-ssl.mzstatic.com/image/thumb/Music114/v4/4c/31/ba/4c31ba09-c78d-897e-6249-f25afe489382/20BMR0002146.rgb.jpg/60x60bb.png",
            attributes: {
              height: "60",
            },
          },
          {
            label:
              "https://is5-ssl.mzstatic.com/image/thumb/Music114/v4/4c/31/ba/4c31ba09-c78d-897e-6249-f25afe489382/20BMR0002146.rgb.jpg/170x170bb.png",
            attributes: {
              height: "170",
            },
          },
        ],
        "im:itemCount": {
          label: "16",
        },
        "im:price": {
          label: "$11.99",
          attributes: {
            amount: "11.99000",
            currency: "USD",
          },
        },
        "im:contentType": {
          "im:contentType": {
            attributes: {
              term: "Album",
              label: "Album",
            },
          },
          attributes: {
            term: "Music",
            label: "Music",
          },
        },
        rights: {
          label: "℗ 2021 Big Machine Label Group, LLC",
        },
        title: {
          label: "Life Rolls On - Florida Georgia Line",
        },
        link: {
          attributes: {
            rel: "alternate",
            type: "text/html",
            href:
              "https://music.apple.com/us/album/life-rolls-on/1541974083?uo=2",
          },
        },
        id: {
          label:
            "https://music.apple.com/us/album/life-rolls-on/1541974083?uo=2",
          attributes: {
            "im:id": "1541974083",
          },
        },
        "im:artist": {
          label: "Florida Georgia Line",
          attributes: {
            href:
              "https://music.apple.com/us/artist/florida-georgia-line/399241518?uo=2",
          },
        },
        category: {
          attributes: {
            "im:id": "6",
            term: "Country",
            scheme: "https://music.apple.com/us/genre/music-country/id6?uo=2",
            label: "Country",
          },
        },
        "im:releaseDate": {
          label: "2021-02-12T00:00:00-07:00",
          attributes: {
            label: "February 12, 2021",
          },
        },
      },
    ],
  },
};
