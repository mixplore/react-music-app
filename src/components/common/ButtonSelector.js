import React from "react";
import "./ButtonSelector.scss";

export default function ButtonSelector({ buttons, selectedButton, onButtonSelect }) {
  return (
    <div role="radiogroup" className="button-selector">
      {buttons.map((btn) => (
        <div key={btn.id}>
          <input
            id={btn.id}
            className="select-button"
            type="radio"
            name="select-option"
            value={btn.value}
            checked={selectedButton === btn.id}
            onChange={onButtonSelect}
          />
          <label htmlFor={btn.id}>{btn.title}</label>
        </div>
      ))}
    </div>
  );
}
