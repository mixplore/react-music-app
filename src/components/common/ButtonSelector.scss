.button-selector {
  display: flex;
  flex-direction: row;

  input[type="radio"] {
    visibility: hidden;
  }

  .select-button {
    + label {
      width: 6rem;
      height: 5rem;
      padding: 0.3rem;
      margin: 0 1rem 1rem 0;
      border-radius: 0.2rem;
      display: flex;
      align-items: center;
      justify-content: center;
      color: #fff;
      background-color: #d3d3d3;
      box-shadow: 0 2px 4px 0 rgba(0, 0, 0, 0.1);
      cursor: pointer;
    }

    + label:hover {
      box-shadow: 0 8px 16px 0 rgba(0, 0, 0, 0.2);
    }

    &:checked + label {
      background-color: #791abe;
    }
  }
}
