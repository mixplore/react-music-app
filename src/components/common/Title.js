import React from "react";
import "./Title.scss";

export default function Title({ text }) {
  return (
    <h2 className="title">
      {text}
    </h2>
  );
}
