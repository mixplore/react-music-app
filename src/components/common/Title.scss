.title {
    color: #000;
    text-transform: capitalize;
}