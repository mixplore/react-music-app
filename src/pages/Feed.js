import React, { useCallback, useMemo, useState } from "react";
import { withRouter } from "react-router-dom";
import "./Feed.scss";
import Title from "../components/common/Title";
import ChartContainer from "../components/common/ChartContainer";
import ChartRow from "../components/common/ChartRow";
import ChartCell from "../components/common/ChartCell";
import Image from "../components/common/Image";
import ButtonSelector from "../components/common/ButtonSelector";
import SearchBar from "../components/common/SearchBar";
import useFeed from "../custom-hooks/useFeed";
import FavouriteIcon from "../components/common/FavouriteIcon";

function Feed({ history }) {
  const {
    feedData,
    switchFeedType,
    isLoading,
    feedType,
    setFavouriteRecord,
  } = useFeed();
  const [searchValue, setSearchValue] = useState(null);

  const selectButtons = [
    {
      id: "topalbums",
      title: "Top 100 Albums",
      value: "topalbums",
    },
    {
      id: "topsongs",
      title: "Top 100 Songs",
      value: "topsongs",
    },
  ];

  const handleFeedSelection = useCallback((event) => {
    switchFeedType(event.target.value);
    history.push(`/${event.target.value}`);
  }, []);

  const handleSearch = useCallback(([, record]) => {
    if (searchValue === null) return record;
    else if (
      (record &&
        record.title.toLowerCase().includes(searchValue.toLowerCase())) ||
      (record &&
        record.albumTitle &&
        record.albumTitle.toLowerCase().includes(searchValue.toLowerCase())) ||
      (record &&
        record.artist.toLowerCase().includes(searchValue.toLowerCase()))
    ) {
      return record;
    }
  });

  const renderChartRow = useMemo(
    () => (record) => (
      <ChartRow key={record.id}>
        <ChartCell>
          <FavouriteIcon isFavourite={record.isFavourite} onSetFavourite={() => setFavouriteRecord(record.id)}/>
          {record.place}
        </ChartCell>
        <ChartCell>
          <Image
            src={record.coverImage.src}
            height={record.coverImage.height}
          />
        </ChartCell>
        <ChartCell>{record.title}</ChartCell>
        <ChartCell>{record.artist}</ChartCell>
        <ChartCell>{record.genre}</ChartCell>
        {record.numberOfTracks && (
          <ChartCell>{record.numberOfTracks}</ChartCell>
        )}
        {record.albumTitle && <ChartCell>{record.albumTitle}</ChartCell>}
        <ChartCell>{record.price}</ChartCell>
        <ChartCell>{record.releaseDate}</ChartCell>
      </ChartRow>
    ),
    [feedData]
  );

  const renderChartHeader = useMemo(
    () => (
      <ChartRow isHeader>
        <ChartCell>#</ChartCell>
        <ChartCell>Cover</ChartCell>
        <ChartCell>Title</ChartCell>
        <ChartCell>Artist</ChartCell>
        <ChartCell>Genre</ChartCell>
        <ChartCell>
          {feedType === "topalbums" ? `Tracks No.` : `Album`}
        </ChartCell>
        <ChartCell>Price</ChartCell>
        <ChartCell>Release Date</ChartCell>
      </ChartRow>
    ),
    [feedType]
  );

  return (
    <div className="page-container">
      <Title text="Aula music tops" />
      <ButtonSelector
        buttons={selectButtons}
        selectedButton={feedType}
        onButtonSelect={handleFeedSelection}
      />
      <SearchBar onValueChanged={(value) => setSearchValue(value)} />
      {isLoading ? (
        <div>Loading...</div>
      ) : (
        <ChartContainer>
          {renderChartHeader}
          {feedData && Object.entries(feedData).length > 0 ? (
            Object.entries(feedData)
              .sort(([, a], [, b]) => a.place - b.place)
              .filter(handleSearch)
              .map(([id, record]) => renderChartRow(record))
          ) : (
            <div>No items found</div>
          )}
        </ChartContainer>
      )}
    </div>
  );
}

export default withRouter(Feed);
