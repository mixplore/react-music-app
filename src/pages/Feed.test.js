import React from "react";
import renderer from "react-test-renderer";
import { createMemoryHistory } from "history";
import { Router } from "react-router-dom";
import { mount, act } from "enzyme";
import Feed from "./Feed";
import { FeedProvider } from "../FeedContext";
import Title from "../components/common/Title";
import ButtonSelector from "../components/common/ButtonSelector";
import mockAxios from "axios";
import topAlbumsMock from "../__mocks__/topAlbumsMock";

describe(`#render`, () => {
  beforeAll(() => {
    const mockResponse = {
      feed: {
        entry: topAlbumsMock,
      },
    };
    mockAxios.get.mockImplementationOnce(() =>
      Promise.resolve({
        data: mockResponse,
      })
    );
  });
  
  const history = createMemoryHistory();
  test(`Snapshot test - Feed page is rendered correctly`, () => {
    const component = renderer.create(
      <Router history={history}>
        <FeedProvider>
          <Feed />
        </FeedProvider>
      </Router>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test(`It should have a title`, () => {
    const wrapper = mount(
      <Router history={history}>
        <FeedProvider>
          <Feed />
        </FeedProvider>
      </Router>
    );
    const title = wrapper.find(Title);
    expect(title).toHaveLength(1);
    expect(title.text()).toContain("Aula music tops");
  });

  test(`It should render a ButtonSelector component`, () => {
    const wrapper = mount(
      <Router history={history}>
        <FeedProvider>
          <Feed />
        </FeedProvider>
      </Router>
    );
    expect(wrapper.find(ButtonSelector)).toBeDefined();
  });

  test(`It should render a search bar with specified placeholder`, () => {
    const wrapper = mount(
      <Router history={history}>
        <FeedProvider>
          <Feed />
        </FeedProvider>
      </Router>
    );
    const searchBar = wrapper.find(".search-bar");
    expect(searchBar).toBeDefined();
    expect(searchBar.children().find("input")).toBeDefined();
    expect(searchBar.children().find("input").props().placeholder).toEqual(
      "Search for..."
    );
  });
});
