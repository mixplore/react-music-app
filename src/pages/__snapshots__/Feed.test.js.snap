// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#render Snapshot test - Feed page is rendered correctly 1`] = `
<div
  className="page-container"
>
  <h2
    className="title"
  >
    Aula music tops
  </h2>
  <div
    className="button-selector"
    role="radiogroup"
  >
    <div>
      <input
        checked={true}
        className="select-button"
        id="topalbums"
        name="select-option"
        onChange={[Function]}
        type="radio"
        value="topalbums"
      />
      <label
        htmlFor="topalbums"
      >
        Top 100 Albums
      </label>
    </div>
    <div>
      <input
        checked={false}
        className="select-button"
        id="topsongs"
        name="select-option"
        onChange={[Function]}
        type="radio"
        value="topsongs"
      />
      <label
        htmlFor="topsongs"
      >
        Top 100 Songs
      </label>
    </div>
  </div>
  <div
    className="search-bar"
  >
    <input
      onChange={[Function]}
      placeholder="Search for..."
      type="text"
    />
  </div>
  <ul
    className="chart-container"
  >
    <li
      className="row header"
    >
      <div
        className="cell"
      >
        #
      </div>
      <div
        className="cell"
      >
        Cover
      </div>
      <div
        className="cell"
      >
        Title
      </div>
      <div
        className="cell"
      >
        Artist
      </div>
      <div
        className="cell"
      >
        Genre
      </div>
      <div
        className="cell"
      >
        Tracks No.
      </div>
      <div
        className="cell"
      >
        Price
      </div>
      <div
        className="cell"
      >
        Release Date
      </div>
    </li>
    <div>
      No items found
    </div>
  </ul>
</div>
`;
